### PR TITLE
[SPIKE][DO NOT MERGE] Create IE-compatible configuration merging function

### PR DESCRIPTION
**Pull request by querkmachine**
Spike to see if I could make a version of the configuration merging function that works back to IE8. 

This exists as a helper function (`getModuleConfig`) that takes an unlimited number of configuration objects, flattens them into a non-nested list of key-value pairs, and merges them into one object. Objects are merged from left to right, in the order they're passed into the function (aka, the last object has the highest priority). 

Also includes a polyfill for `Element.prototype.dataset`. This is derived from [the one formerly served by polyfill.io](https://raw.githubusercontent.com/Financial-Times/polyfill-library/13cf7c340974d128d557580b5e2dafcd1b1192d1/polyfills/Element/prototype/dataset/polyfill.js), with some modifications:
- The regex used to detect data-* attributes has been altered so that it catches attributes containing full stops (.), as we're using these to namespace i18n attributes.
- A feature detection for `Object.prototype.__defineGetter__` and `Object.prototype.__defineSetter__` has been added, as these are required by the polyfill but unavailable in IE8, and the existing `Object.defineProperty` polyfill doesn't cover these. Instead, a (very hacky) fallback is used if these aren't detected.

## Seeing it in action

The Accordion component has been modified to accept i18n options and to use the helper function. Currently, no visual changes take place and the resulting configuration object is logged in the console instead (in a way that IE8 won't fall over on!)

A new example has been added to the review app, ['Accordion with localisations'](https://govuk-frontend-pr-2810.herokuapp.com/components/accordion/with-localisations/preview), which overrides the `i18n.showSection` and `i18n.showAllSections` strings, exhibiting how the default configuration and data-* attribute configuration options get merged and flattened.

If all is working as expected, you should get this logged to the console on every browser: 

```
Accordion configuration:
    i18n.hideAllSections: Hide all sections
    i18n.hideSection: Hide <span class="govuk-visually-hidden">this section</span>
    i18n.showAllSections: Show everything
    i18n.showSection: Show<span class="govuk-visually-hidden"> something</span>
    module: govuk-accordion
```

### CODE REVIEW

File: src/govuk/common.mjs [hunk 1]
```diff
@@ -26,3 +30,64 @@ export function generateUniqueID () {
...
+  // one-by-one into configObject. Any duplicate keys will override the existing
+  // key with the new value.
+  for (var i = 0; i < arguments.length; i++) {
+    var obj = flattenObject(arguments[i])
+    for (var key in obj) {
+      if (Object.prototype.hasOwnProperty.call(obj, key)) {
+        configObject[key] = obj[key]
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (querkmachine) [reply src/govuk/common.mjs:90]: To my mind, there isn't much practical difference between allowing an arbitrary number of objects and codifying specific objects, as either way it depends on the component JS passing them in the correct order to obtain the expected result.

We'd then need to combine them into a loopable array anyway. Albeit, not much complexity, probably something like...
```js 
var configArray = [flattenObject(defaultOptions), flattenObject(initOptions), datasetOptions]
for (var i = 0; i < configArray.length; i++) { 
  var obj = configArray[i]
  // …
}
```

The main downside I see of this approach is that, if config objects are passed in the wrong order, this'll could make a mess—but I think we're probably okay with that given it's an internal API. 

Good points on IDE suggestion and flattening dataset though. I'm not against codifying if we think that makes it easier to understand.

File: src/govuk/components/accordion/accordion.mjs [hunk 2]
```diff
@@ -15,17 +15,35 @@
 
 */
 
-import { nodeListForEach } from '../../common.mjs'
+import { nodeListForEach, getModuleConfig } from '../../common.mjs'
 import '../../vendor/polyfills/Function/prototype/bind.mjs'
 import '../../vendor/polyfills/Element/prototype/classList.mjs'
 
-function Accordion ($module) {
+function Accordion ($module, options) {
```
[the remark this answers is not in the document]
Comment (querkmachine) [reply src/govuk/components/accordion/accordion.mjs:22]: I'd love to, but IE doesn't support this syntax!

[the remark this answers is not in the document]
Comment (36degrees) [reply src/govuk/components/accordion/accordion.mjs:22]: Of course 🤦🏻

File: src/govuk/common.mjs [hunk 3]
```diff
@@ -26,3 +30,64 @@ export function generateUniqueID () {
...
+
+/**
+ * Config flattening function. Takes any number of objects, flattens them into
+ * namespaced key-value pairs, (e.g. {'i18n.showSection': 'Show section'}) and
+ * combines them together, with greatest priority on the LAST item passed in.
+ *
+ * @param {...object} - Any number of objects to merge together.
+ * @returns {object} - A flattened object of key-value pairs.
+ */
+export function getModuleConfig () {
```
[the remark this answers is not in the document]
Comment (querkmachine) [reply src/govuk/common.mjs:42]: Sounds good to me 👍

File: package/govuk-esm/components/accordion/accordion.mjs [hunk 4]
```diff
@@ -15,16 +15,27 @@
 
 */
 
-import { nodeListForEach } from '../../common.mjs'
+import { nodeListForEach, getModuleConfig } from '../../common.mjs'
 import '../../vendor/polyfills/Function/prototype/bind.mjs'
 import '../../vendor/polyfills/Element/prototype/classList.mjs'
 
-function Accordion ($module) {
+function Accordion ($module, options) {
```
[the remark this answers is not in the document]
Comment (querkmachine) [reply package/govuk-esm/components/accordion/accordion.mjs:22]: Yeah, we should probably standardise on `config`—it feels a bit more specific.

On object vs. init, they feel like different cases to me. There isn't any other way of passing config into `initAll`, so we're kind of stuck with what we have there. 

For components, doing it on object creation makes more sense to me. Our existing use of `init` functions seems like more of a legacy layover from not being able to use native constructors. (I think we've even briefly discussed refactoring out the `init`s after we drop IE JS support.)

File: src/govuk/common.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (36degrees) [reply src/govuk/common.mjs]: Picking up a conversation we've been having on Slack, we've talked about a potential move to having a 'base component' that would take care of setting up config etc that each component can extend as part of the JS work in 5.0.

On the assumption that we are going to go down that route in the future, and therefore we have a future state where the merging of the different configs happens in a single place, I think we can go ahead with the current approach and just consider this as a sort of tech debt.

File: package/govuk-esm/components/accordion/accordion.mjs (hunk 4 above)
[the remark this answers is not in the document]
Comment (36degrees) [reply package/govuk-esm/components/accordion/accordion.mjs:22]: Makes sense to me 👍🏻

**Comment by querkmachine** [comment]
The team seems to be happy with the state of this. Will close this PR for now. This branch could serve as a starting point for #2699 when we get to that.

**Pull request closed without merging**